File: frontend/src/lib/types/study-plan.ts
/**
 * Study Plan types.
 * AI-generated personalized study schedules.
 */

export type PlanStatus = 'active' | 'completed' | 'paused' | 'cancelled';
export type TopicDifficulty = 'easy' | 'medium' | 'hard';
export type TopicStatus = 'not_started' | 'in_progress' | 'completed' | 'skipped';
export type SessionType = 'learn' | 'review' | 'practice' | 'quiz';
export type SessionStatus = 'scheduled' | 'in_progress' | 'completed' | 'skipped' | 'rescheduled';
export type AdjustmentType = 'reschedule' | 'add_review' | 'extend_deadline' | 'increase_hours' | 'reduce_scope';
export type AdjustmentStatus = 'pending' | 'accepted' | 'rejected';

export interface StudyPlan {
  id: string;
  notebook_id: string;
  title: string;
  description?: string;
  deadline: string;
  available_hours_per_day: number;
  total_study_hours: number;
  status: PlanStatus;
  progress_percentage: number;
  created_at: string;
  updated_at: string;
}

export interface StudyPlanCreate {
  notebook_id: string;
  title: string;
  description?: string;
  deadline: string;
  available_hours_per_day?: number;
}

export interface StudyPlanUpdate {
  title?: string;
  description?: string;
  deadline?: string;
  available_hours_per_day?: number;
  status?: PlanStatus;
}

export interface StudyTopic {
  id: string;
  plan_id: string;
  name: string;
  description?: string;
  difficulty: TopicDifficulty;
  estimated_hours: number;
  priority: number;
  source_ids: string[];
  prerequisites: string[];
  status: TopicStatus;
  mastery_level: number;
  created_at: string;
}

export interface StudyTopicCreate {
  plan_id: string;
  name: string;
  description?: string;
  difficulty?: TopicDifficulty;
  estimated_hours: number;
  priority?: number;
  source_ids?: string[];
  prerequisites?: string[];
}

export interface StudyTopicUpdate {
  name?: string;
  description?: string;
  difficulty?: TopicDifficulty;
  estimated_hours?: number;
  priority?: number;
  status?: TopicStatus;
  mastery_level?: number;
}

export interface StudySession {
  id: string;
  plan_id: string;
  topic_id: string;
  scheduled_date: string;
  scheduled_duration_minutes: number;
  actual_start?: string;
  actual_end?: string;
  session_type: SessionType;
  status: SessionStatus;
  notes?: string;
  rating?: number;
  created_at: string;
  topic_name?: string;
}

export interface StudySessionCreate {
  plan_id: string;
  topic_id: string;
  scheduled_date: string;
  scheduled_duration_minutes: number;
  session_type?: SessionType;
}

export interface StudySessionUpdate {
  scheduled_date?: string;
  scheduled_duration_minutes?: number;
  session_type?: SessionType;
  actual_start?: string;
  actual_end?: string;
  status?: SessionStatus;
  notes?: string;
  rating?: number;
}

export interface PlanAdjustment {
  id: string;
  plan_id: string;
  adjustment_type: AdjustmentType;
  reason: string;
  original_value?: string;
  suggested_value?: string;
  status: AdjustmentStatus;
  created_at: string;
}

export interface StudyPlanFull extends StudyPlan {
  topics: StudyTopic[];
  sessions: StudySession[];
  adjustments: PlanAdjustment[];
  days_remaining: number;
  completed_hours: number;
  upcoming_sessions: StudySession[];
}

export interface DailySchedule {
  date: string;
  sessions: StudySession[];
  total_hours: number;
  is_today: boolean;
}

export interface WeeklySchedule {
  plan_id: string;
  week_start: string;
  week_end: string;
  days: DailySchedule[];
  total_planned_hours: number;
  total_completed_hours: number;
}

export interface StudyPlanStats {
  plan_id: string;
  total_topics: number;
  completed_topics: number;
  total_sessions: number;
  completed_sessions: number;
  total_planned_hours: number;
  total_completed_hours: number;
  average_rating?: number;
  on_track: boolean;
  days_remaining: number;
  hours_per_day_needed: number;
}

export interface PlanGenerationRequest {
  notebook_id: string;
  title: string;
  description?: string;
  deadline: string;
  available_hours_per_day?: number;
  include_reviews?: boolean;
  include_practice?: boolean;
  focus_areas?: string[];
}

export interface PlanGenerationResult {
  plan: StudyPlan;
  topics: StudyTopic[];
  sessions: StudySession[];
  total_hours: number;
  days_with_sessions: number;
  warnings: string[];
}

File: open_notebook/domain/study_plan.py
"""
Study Plan domain models.
AI-generated personalized study schedules based on content and deadlines.
"""

from datetime import datetime
from enum import Enum
from typing import Optional, List
from pydantic import BaseModel, Field


class PlanStatus(str, Enum):
    """Study plan status."""
    ACTIVE = "active"
    COMPLETED = "completed"
    PAUSED = "paused"
    CANCELLED = "cancelled"


class TopicDifficulty(str, Enum):
    """Topic difficulty level."""
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


class TopicStatus(str, Enum):
    """Study topic status."""
    NOT_STARTED = "not_started"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    SKIPPED = "skipped"


class SessionType(str, Enum):
    """Study session type."""
    LEARN = "learn"
    REVIEW = "review"
    PRACTICE = "practice"
    QUIZ = "quiz"


class SessionStatus(str, Enum):
    """Study session status."""
    SCHEDULED = "scheduled"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    SKIPPED = "skipped"
    RESCHEDULED = "rescheduled"


class AdjustmentType(str, Enum):
    """Plan adjustment type."""
    RESCHEDULE = "reschedule"
    ADD_REVIEW = "add_review"
    EXTEND_DEADLINE = "extend_deadline"
    INCREASE_HOURS = "increase_hours"
    REDUCE_SCOPE = "reduce_scope"


class AdjustmentStatus(str, Enum):
    """Plan adjustment status."""
    PENDING = "pending"
    ACCEPTED = "accepted"
    REJECTED = "rejected"


# Request/Response models

class StudyPlanCreate(BaseModel):
    """Create a new study plan."""
    notebook_id: str
    title: str
    description: Optional[str] = None
    deadline: datetime
    available_hours_per_day: float = Field(default=2.0, ge=0.5, le=12.0)


class StudyPlanUpdate(BaseModel):
    """Update study plan."""
    title: Optional[str] = None
    description: Optional[str] = None
    deadline: Optional[datetime] = None
    available_hours_per_day: Optional[float] = Field(default=None, ge=0.5, le=12.0)
    status: Optional[PlanStatus] = None


class StudyPlan(BaseModel):
    """Study plan model."""
    id: str
    notebook_id: str
    title: str
    description: Optional[str] = None
    deadline: datetime
    available_hours_per_day: float = 2.0
    total_study_hours: float = 0.0
    status: PlanStatus = PlanStatus.ACTIVE
    progress_percentage: float = 0.0
    created_at: datetime
    updated_at: datetime


class StudyTopicCreate(BaseModel):
    """Create a study topic."""
    plan_id: str
    name: str
    description: Optional[str] = None
    difficulty: TopicDifficulty = TopicDifficulty.MEDIUM
    estimated_hours: float = Field(ge=0.25)
    priority: int = Field(default=5, ge=1, le=10)
    source_ids: List[str] = Field(default_factory=list)
    prerequisites: List[str] = Field(default_factory=list)


class StudyTopicUpdate(BaseModel):
    """Update study topic."""
    name: Optional[str] = None
    description: Optional[str] = None
    difficulty: Optional[TopicDifficulty] = None
    estimated_hours: Optional[float] = Field(default=None, ge=0.25)
    priority: Optional[int] = Field(default=None, ge=1, le=10)
    status: Optional[TopicStatus] = None
    mastery_level: Optional[float] = Field(default=None, ge=0.0, le=100.0)


class StudyTopic(BaseModel):
    """Study topic model."""
    id: str
    plan_id: str
    name: str
    description: Optional[str] = None
    difficulty: TopicDifficulty = TopicDifficulty.MEDIUM
    estimated_hours: float
    priority: int = 5
    source_ids: List[str] = Field(default_factory=list)
    prerequisites: List[str] = Field(default_factory=list)
    status: TopicStatus = TopicStatus.NOT_STARTED
    mastery_level: float = 0.0
    created_at: datetime


class StudySessionCreate(BaseModel):
    """Create a study session."""
    plan_id: str
    topic_id: str
    scheduled_date: datetime
    scheduled_duration_minutes: int = Field(ge=15, le=480)
    session_type: SessionType = SessionType.LEARN


class StudySessionUpdate(BaseModel):
    """Update study session."""
    scheduled_date: Optional[datetime] = None
    scheduled_duration_minutes: Optional[int] = Field(default=None, ge=15, le=480)
    session_type: Optional[SessionType] = None
    actual_start: Optional[datetime] = None
    actual_end: Optional[datetime] = None
    status: Optional[SessionStatus] = None
    notes: Optional[str] = None
    rating: Optional[int] = Field(default=None, ge=1, le=5)


class StudySession(BaseModel):
    """Study session model."""
    id: str
    plan_id: str
    topic_id: str
    scheduled_date: datetime
    scheduled_duration_minutes: int
    actual_start: Optional[datetime] = None
    actual_end: Optional[datetime] = None
    session_type: SessionType = SessionType.LEARN
    status: SessionStatus = SessionStatus.SCHEDULED
    notes: Optional[str] = None
    rating: Optional[int] = None
    created_at: datetime
    # Populated when joined
    topic_name: Optional[str] = None


class PlanAdjustment(BaseModel):
    """Plan adjustment suggestion."""
    id: str
    plan_id: str
    adjustment_type: AdjustmentType
    reason: str
    original_value: Optional[str] = None
    suggested_value: Optional[str] = None
    status: AdjustmentStatus = AdjustmentStatus.PENDING
    created_at: datetime


class PlanAdjustmentResponse(BaseModel):
    """Response to a plan adjustment."""
    adjustment_id: str
    accepted: bool


# Response models with nested data

class StudyPlanWithTopics(StudyPlan):
    """Study plan with topics."""
    topics: List[StudyTopic] = Field(default_factory=list)


class StudyPlanFull(StudyPlanWithTopics):
    """Complete study plan with all data."""
    sessions: List[StudySession] = Field(default_factory=list)
    adjustments: List[PlanAdjustment] = Field(default_factory=list)
    days_remaining: int = 0
    completed_hours: float = 0.0
    upcoming_sessions: List[StudySession] = Field(default_factory=list)


class DailySchedule(BaseModel):
    """Schedule for a single day."""
    date: datetime
    sessions: List[StudySession] = Field(default_factory=list)
    total_hours: float = 0.0
    is_today: bool = False


class WeeklySchedule(BaseModel):
    """Weekly schedule view."""
    plan_id: str
    week_start: datetime
    week_end: datetime
    days: List[DailySchedule] = Field(default_factory=list)
    total_planned_hours: float = 0.0
    total_completed_hours: float = 0.0


class StudyPlanStats(BaseModel):
    """Statistics for study plan."""
    plan_id: str
    total_topics: int = 0
    completed_topics: int = 0
    total_sessions: int = 0
    completed_sessions: int = 0
    total_planned_hours: float = 0.0
    total_completed_hours: float = 0.0
    average_rating: Optional[float] = None
    on_track: bool = True
    days_remaining: int = 0
    hours_per_day_needed: float = 0.0


class TopicAnalysis(BaseModel):
    """AI-extracted topic from sources."""
    name: str
    description: str
    difficulty: TopicDifficulty
    estimated_hours: float
    priority: int
    source_ids: List[str]
    prerequisites: List[str] = Field(default_factory=list)
    key_concepts: List[str] = Field(default_factory=list)


class PlanGenerationRequest(BaseModel):
    """Request to generate a study plan."""
    notebook_id: str
    title: str
    description: Optional[str] = None
    deadline: datetime
    available_hours_per_day: float = Field(default=2.0, ge=0.5, le=12.0)
    include_reviews: bool = True
    include_practice: bool = True
    focus_areas: Optional[List[str]] = None  # Optional list of topics to prioritize


class PlanGenerationResult(BaseModel):
    """Result of plan generation."""
    plan: StudyPlan
    topics: List[StudyTopic]
    sessions: List[StudySession]
    total_hours: float
    days_with_sessions: int
    warnings: List[str] = Field(default_factory=list)
